text/0000-remove-shell-processes.md

```diff
@@ -0,0 +1,401 @@
+
+### Example 1 - A Direct process
+The follow direct commands:
+```
+docker run --entrypoint launcher <image> -- env
+docker run --entrypoint launcher <image> -- echo hello '$WORLD' 
+```
+will become the following, using the new platform API
+```
+docker run --entrypoint launcher <image> env
+docker run --entrypoint launcher <image> echo hello '${WORLD}'
+```
+Previously, in the second command in this example `${WORLD}` would not have been interpolated because this is a direct process; instead the output would include the literal string `$WORLD`. With the changes proposed, `${WORLD}` will now be evaluated, even though the process is direct.
+
+### Example 2 - A Shell Process
+The follow custom shell command:
+```
+docker run --entrypoint launcher <image> echo hello '${WORLD}'
+docker run --entrypoint launcher <image> echo hello '${WORLD:-world}'
+```
+will become the following, using the new platform API
+```
+docker run --entrypoint launcher <image> echo hello '${WORLD}'
+docker run --entrypoint launcher <image> bash -c 'echo hello "${WORLD:-world}"'
+```
+The first command in this example did not need to change to behave as expected with the new API. Previously it was necessary to use a shell process in order to evaluate `${WORLD}`. Now, the shell is unescessary. However, if the user wants to use more features of Bash interpolation they will need to explicitly invoke a shell, as shown in the second command in this example.
+
+### Example 3 - A Script Process
+The follow custom script command:
+```
+docker run --entrypoint launcher <image> 'for opt in $JAVA_OPTS; do echo $opt; done'
+```
+will become the following, using the new platform API
+```
+docker run --entrypoint launcher <image> bash -c 'for opt in $JAVA_OPTS; do echo $opt; done'
+```
+
+## What About Profile Scripts?
+
+### `<layer>/profile.d/*`
+When a buildpack upgrades to the new buildpack API, buildpack-provided profile scripts will no longer be supported. Instead buildpacks can use `exec.d`. Most existing profiles scripts can be easily converted.
+
+**Example 1: `jkutner/sshd` **
+
+The following profile script example was taken from https://github.com/jkutner/sshd-buildpack/blob/master/sbin/sshd.sh
+```
+#!/usr/bin/env bash
+
+if [[ "${SSH_DISABLED:-}" != "true" ]]; then
+  ssh_layer=$(realpath $(dirname ${BASH_SOURCE[0]})/..)
+  ssh_dir=$(realpath $HOME/.ssh)
+
+  mkdir -p $ssh_dir
+  cat $ssh_layer/id_rsa.pub >> $ssh_dir/authorized_keys
+
+  cat << EOF >> $ssh_dir/sshd_config
+HostKey $ssh_layer/id_rsa
+AuthorizedKeysFile $ssh_dir/authorized_keys
+EOF
+
+  chmod 600 /workspace/.ssh/*
+
+  ssh_port=${SSH_PORT:-"2222"}
+  echo "at=sshd state=starting user=$(whoami) port=${ssh_port}"
+  /usr/sbin/sshd -f $ssh_dir/sshd_config -o "Port ${ssh_port}"
+fi
+```
+This script does not require any changes and would work as an `exec.d` or `profile.d`.
```

Comment:
I'd still need to move the script in the `exec.d` though right? are there any restrictions like filename or permissions that might change? (I'm sure these are all fine but would like to capture)

Reply:
Good point. We should capture two things:
1. The script must move to `<layer>/exec.d/`
2. The runtime user must have permission to execute this file (e.g. executable bit should be set).

Reply:
Added in https://github.com/buildpacks/rfcs/pull/168/commits/b49861f13dd11b080099b0da7ac8b49e9611f113